Add v1.0.0 changelogs (#39)
meilisearch/engine-team: meilisearch-release-changelog/v1.0.0.md

# v1.0.0 release changelogs

<!-- The following line should NOT be put in the official release changelogs -->
⚠️ Since this is a release candidate (RC), we do NOT recommend using it in a production environment. Is something not working as expected? We welcome [bug reports](https://github.com/meilisearch/meilisearch/issues/new/choose) and [feedback about new features](https://github.com/meilisearch/product/discussions).

> We are proud to announce we are releasing the first major release of Meilisearch! 🎉 This new release, contrary to v0.30.0, focus on stabilization instead of bringing new features. We took the opportunity to make the final touch regarding breaking changes in the CLI and the error handler.

Author: @maryamsulemani97, I was thinking, I don't feel really happy with my original sentence. Could we make it more "positive"? For example, precise breaking changes are here for the best. Also, maybe should I talk about the new language support?
I will try to do a better suggestion.

> We are proud to announce we are releasing the first major release of Meilisearch! 🎉 This new release, contrary to v0.30.0, focus on stabilization instead of bringing new features. We took the opportunity to make the final touch regarding breaking changes in the CLI and the error handler.

Author: ```suggestion
We are proud to announce we are releasing the first major release of Meilisearch! 🎉

This new release focuses on stabilizing the CLI instead of only introducing new features. We took this opportunity to introduce the final breaking changes to the CLI and the error handler which would make the user experience more intuitive.

Despite our focus on the Meilisearch stabilization, we also took the time to improve the Meilisearch support of the Japanese and Chinese languages!
```

> We are proud to announce we are releasing the first major release of Meilisearch! 🎉 This new release, contrary to v0.30.0, focus on stabilization instead of bringing new features. We took the opportunity to make the final touch regarding breaking changes in the CLI and the error handler.

Reviewer: ```suggestion
We are proud to announce the first major release of Meilisearch! 🎉
	
This new release focuses on stabilizing the CLI instead of only introducing new features. We took this opportunity to introduce final breaking changes to the CLI and error handler to enhance user experience.
	
Despite our focus on stabilizing Meilisearch, we also took the time to improve support for Japanese and Chinese!

Since Meilisearch is now stable, we will not introduce any new breaking changes in the future versions of Meilisearch until releasing v2.0.0. Here you have more information about what we consider [as breaking for Meilisearch](https://github.com/meilisearch/engine-team/blob/main/resources/versioning-policy.md).

<!-- The following lines should ONLY be put in the official release changelogs -->

> 🧰 Most integrations (SDKs, clients, libraries) will be compatible with this version within four hours of the release. Sometimes this can take up to 48 hours, depending on the issues we encounter during the release.

Author: Already reviewed last time by docs team, I copied pasted the previous release 😉

<!-- The following lines should ONLY be put in the official release changelogs -->

> Here is the exhaustive list of integrations  **not immediately compatible** with v1.0.0:

Author: Same here

- TBD

<!-- The following line should NOT be put in the official release changelogs -->
📖 Meilisearch's official documentation will be ready for the official v1.0.0 release. Meanwhile, work-in-progress documentation is available [here]().

# New features and improvements 🔥

# Charabia

TBD

# Generate master key option

You can now activate the `--generate-master-key` CLI option when launching Meilisearch.

> This option is particular since it won't launch Meilisearch but will generate a safe-to-use master key for Meilisearch and will exit without any error.

Reviewer 2: ```suggestion
This option generates a master key that is safe to use with Meilisearch and then immediately exit.
```

> This option is particular since it will not launch Meilisearhc but will generate a safe-to-use master key for Meilisearch and will exit without any error.

Reviewer 3: ```suggestion
This option is particular since it won't launch Meilisearch but will generate a safe-to-use master key for Meilisearch and will exit without any error.
```

Done in #3274 by @dureuill

## Other improvements

> * Re-enable importing from dumps v1: you can now migrate from Meilisearch v0.20.0 and older to the latest version using a dump (#3170) @dureuill

Reviewer 3: ```suggestion
* Re-enable importing from dumps v1: you can now migrate from Meilisearch v0.15.0 and older to the latest version using a dump (#3170) @dureuill
```

From what I see, the dump v1 have been released in the v0.15.0; https://github.com/meilisearch/meilisearch/pull/887

Reviewer 2: Does that mean that we can import from v0.15 and **newer** (new range is v0.15-v0.20)?

* Improve [`download-latest.sh`](https://github.com/meilisearch/meilisearch/blob/main/download-latest.sh) script: integrate apple-silicon binary (#3169) @curquiza
* Clarify error message when the DB and engine versions are incompatible (#3279) @dureuill
* Search speed improvements:
  * synonyms (TBD)
  * ranking rules (TDB)
  * ... (TBD)

# Breaking changes ⚠️

Since we are releasing v1.0.0, this is the last time we are introducing breaking changes to this release before v2.0.0.

## New CLI behavior

* For security reasons, Meilisearch will now reject master keys that are less than 16 bytes (#3274) @dureuill
* Remove `--max-index-size` and `--max-task-db` configuration options (#3278) @dureuill

> Indeed, these options did not include the real space Meilisearch takes on the disk and were not adapted to the user's needs.

Author: Docs team will help us

> Indeed, these options did not include the real space Meilisearch takes on the disk and were not adapted to the user's needs.

Reviewer 2: I wonder if we should maybe link to the [tradeoff discussion](https://github.com/meilisearch/meilisearch/discussions/3280).

Maybe...

```suggestion
These options were not serving their intended purpose of limiting the disk space taken by Meilisearch, for the reasons laid out [here](https://github.com/meilisearch/product/discussions/567#discussioncomment-4322126), but actually articulate a [tradeoff](https://github.com/meilisearch/meilisearch/discussions/3280) between the number of indexes and their maximum size that we'd like to express differently in the future.

In v1, the number of indexes is limited to around 200 under Linux/macOS and around 20 under Windows, with a maximum size of 500GiB per index. We plan to lift these limitations in future versions, please [reach out](https://github.com/meilisearch/meilisearch/discussions/3280) if the new limits impede your usage.
```

Some of this might be out of place in the changelog 🤔... in any case, I'd like this communicated to the users.

> Indeed, these options did not include the real space Meilisearch takes on the disk and were not adapted to the user's needs.

Author: Hum really complex to me, and not really interesting for the users reading the changelog. The docs are more appropriate I think. Maybe a link to the docs could help?

However, maybe we could have some words if some people were using it and what they could do instead.

> Indeed, these options did not include the real space Meilisearch takes on the disk and were not adapted to the user's needs.

Reviewer 2: > However, maybe we could have some words if some people were using it and what they could do instead.

Not much. Reach out to us or limit disk space from their infrastructure.

> Indeed, these options did not include the real space Meilisearch takes on the disk and were not adapted to the user's needs.

Reviewer: I think we can go with @dureuill's suggestion, but maybe shorten it to: 
```suggestion
These options were not serving their intended purpose of limiting the disk space taken by Meilisearch for the reasons laid out [here](https://github.com/meilisearch/product/discussions/567#discussioncomment-4322126). 
Please [reach out](https://github.com/meilisearch/meilisearch/discussions/3280) if the [new limits](add link) impede your usage.
```

* Remove --disable-auto-batching CLI option (#3296) @loiclec
This option was introduced in case of bug, and is not useful anymore since the feature is mature enough.
* Rename dump command from `--dumps-dir` to `--dump-dir` (#3175) @dureuill
* Remove `--snapshot-interval-sec` and `--schedule-snapshot` now accepts an optional integer value specifying the interval in seconds (#3281) @dureuill
* The `--log-level` option does not take any string to filter the logs anymore, but only accepts the following arguments: `ERROR`, `WARN`, `INFO`, `DEBUG`, `TRACE` and `OFF` (#3293) @loiclec

> ## New primary key inference

Reviewer 4: ![cool_kid_pc](https://user-images.githubusercontent.com/3692335/210443023-6d1b451d-675f-4d00-bd8d-abba5405cc3c.jpeg)

When receiving the documents for the first time, and if the `primaryKey` parameter is not present in the request, Meilisearch tries to infer the primary key.

> In this case, Meilisearch will first try to find any key in the document containing `id`, like `puid` or `_id`. If only one key is detected, this key will be considered as the primary key. The changing behavior is about detecting multiple keys containing `id`.

Reviewer 3: ```suggestion
In this case, Meilisearch will first try to find any key in the document ending with `id`, like `puid` or `_id`. If only one key is detected, this key will be considered as the primary key.
```

> In this case, Meilisearch will first try to find any key in the document containing `id`, like `puid` or `_id`. If only one key is detected, this key will be considered as the primary key. The changing behavior is about detecting multiple keys containing `id`.

Reviewer 2: We now look for fields ending with `id`, not simply containing `id`

> Old behavior: if Meilisearch finds multiple keys containing `id`, it will consider the primary key as the first field in the document.

Reviewer 3: ```suggestion
Old behavior: Meilisearch uses the first field in the document containing `id`, ignoring anything else.
```

> New behavior: to respect the JSON specification, if Meilisearch finds multiple keys containing `id`, it will now return an error. Indeed, there is no priority order in JSON documents.

Reviewer 3: I'm not sure we should relate it to the JSON specification, the « real » reason is that meilisearch is probably going to be wrong and need the user to help him :bee: 

```suggestion
New behavior: to respect the JSON specification, if Meilisearch finds multiple keys ending by `id`, it will return an error. Indeed, there is no priority order in JSON documents.
```

If no primary key candidates are detected, Meilisearch will continue to throw an error as before.

> Done in meilisearch/milli#752 and #??? (TBD) by @dureuill

Reviewer 2: ```suggestion
Done in meilisearch/milli#752 and #3269 by @dureuill
```

(the draft PR is already open :-))

> As previously, Meilisearch will throw an error is no key containing `id` are detected and if the `primaryKey` parameter is not present.

Reviewer 3: ```suggestion
As previously, Meilisearch will throw an error if it can't infer any primary key and if the `primaryKey` parameter is not present.
```

## Error handler changes

> * Add a new error `code`: `io_error` (used instead of `internal`) (#3263) @irevoire
> * Add a new error `code`: `too_many_open_files` (used instead of `internal`) (#3263) @irevoire

Reviewer 4: ```suggestion
* Add a new error `type`: `???` (#3263) @irevoire
    * `too_many_open_files` error `code` is linked to that error `type` instead of `internal`
    * `no_space_left_on_device` error `code` is linked to that error `type` instead of `internal`
    *`io_error` error `code` is linked to that error `type` instead of `internal`
```

Reviewer 4: https://meilisearch.slack.com/archives/C04BRF22ZBP/p1672754110902429 Voted here 😃

Author: Waiting for the whole list of @irevoire before changing anything here 😇

* TBD (changes brought by Tamo)

# Fixes 🐞

* Commit SHA and date are now present in the Meilisearch logs when launching Meilisearch from a Docker image (#3212) @brunoocasali

> * Fix `MDB_PAGE_FULL` internal error (#3210) @Kerollmops

Reviewer 2: ```suggestion
* Fix `MDB_PAGE_FULL` internal error that would sometimes occur when inserting documents on macOS (#3210) @Kerollmops
```

To ensure the stability of Meilisearch, we will not introduce any breaking changes in future versions until the release of v2.0.0. Please refer to the following [information](https://github.com/meilisearch/engine-team/blob/main/resources/versioning-policy.md) for a definition of what we consider to be a breaking change in Meilisearch.

<!-- The following lines should ONLY be put in the official release changelogs -->
🧰 Most integrations (SDKs, clients, libraries) will be compatible with this version within four hours of the release. Sometimes this can take up to 48 hours, depending on the issues we encounter during the release.

<!-- The following lines should ONLY be put in the official release changelogs -->
Here is the exhaustive list of integrations  **not immediately compatible** with v1.0.0:
- TBD

<!-- The following line should NOT be put in the official release changelogs -->
📖 Meilisearch's official documentation will be ready for the official v1.0.0 release. Meanwhile, work-in-progress documentation is available [here]().

# New features and improvements 🔥

# Charabia

TBD

# Generate master key option

> You can now take advantage of the `--generate-master-key` CLI option.
> 
> By using that option, you can easily generate a strong master key to secure your Meilisearch instance easily. Using that option won't start Meilisearch and will exit without any error.
> 

Reviewer: ```suggestion
You can now use the `--generate-master-key` CLI option.

It allows you to easily generate a strong master key to secure your Meilisearch instance. Using this option won't start Meilisearch and will exit without any error.

```

Done in #3274 by @dureuill

## Other improvements

* Search speed improvements:
  * Significantly improve the memory usage of search requests containing multiple long words (#3115) @loiclec
  * Significantly improve the performance of the `exactness` ranking rule for search requests containing many words (#3116) @loiclec
  * Translate multi-word synonyms into [phrases](https://docs.meilisearch.com/learn/what_is_meilisearch/features.html#phrase-search) when interpreting a search query. This change improves the relevancy of the results and stabilises the latency of search queries containing many multi-word synonyms, removing a source of denial-of-service attacks. (#3125) @loiclec
  * Improve the performance of the `proximity` ranking rule for search requests ending with short words (#3118) @loiclec

> * You can now migrate from any old versions of Meilisearch (supporting the dump feature) to the latest version using a single dump. We hope to facilitate your migration to v1.0.0 so you can enjoy the latest features as easily as possible (#3170) @dureuill

Reviewer: ```suggestion
* Using a single dump, you can now migrate from any old version of Meilisearch (supporting the dump feature) to the latest version. We hope to facilitate your migration to v1.0.0 to make it as smooth as possible (#3170) @dureuill
```

* Improve [`download-latest.sh`](https://github.com/meilisearch/meilisearch/blob/main/download-latest.sh) script: integrate apple-silicon binary (#3169) @curquiza
* Clarify error message when the DB and engine versions are incompatible (#3279) @dureuill

# Breaking changes ⚠️

Since we are releasing v1.0.0, this is the last time we are introducing breaking changes to this release before v2.0.0.

## New CLI behavior

* For security reasons, Meilisearch will now reject master keys that are less than 16 bytes (#3274) @dureuill
* Remove `--max-index-size` and `--max-task-db` configuration options (#3278) @dureuill
Indeed, these options did not include the real space Meilisearch takes on the disk and were not adapted to the user's needs.

> * Remove `--disable-auto-batching` CLI option (#3296) @loiclec

Reviewer: When talking about instance options, should we mention both the environment variable and CLI option (like line57) or just the CLI option?

Author: Hum yes you're right, would be clearer!

This option was initially introduced as a workaround in case of a bug and is no longer useful since the feature has matured.
* Rename dump command from `--dumps-dir` to `--dump-dir` (#3175) @dureuill

> * Remove `--snapshot-interval-sec` and `--schedule-snapshot` now accepts an optional integer value specifying the interval in seconds (#3281) @dureuill

Reviewer: ```suggestion
* Remove `--snapshot-interval-sec`. `--schedule-snapshot` now accepts an optional integer value specifying the interval in seconds (#3281) @dureuill
```

> * The `--log-level` option and `MEILI_LOG_LEVEL` environment variable now only accept the following arguments: `ERROR`, `WARN`, `INFO`, `DEBUG`, `TRACE` and `OFF` (#3293) @loiclec

Reviewer: ```suggestion
* The `--log-level` option and `MEILI_LOG_LEVEL` environment variable accept the following arguments: `ERROR`, `WARN`, `INFO`, `DEBUG`, `TRACE` and `OFF` (#3293) @loiclec
```

Author: It will change the phrase sentence, which is not what we want to say. Currently, Meilisearch can accept these values, but also any value, like strings, that Meilisearch can interpret to apply a filter to the logs.
We did not code anything, we only sent the string to `logger` aka the Rust library that handles the logs and which interprets the string.
We changed this, Meilisearch now only does what the documentation details, so Meilisearch now only accepts the following value (`ERROR`...)

Here is [the related issue](https://github.com/meilisearch/meilisearch/issues/3292) (no team are pinged because it does not impact anyone, except the users who made tricky and un-documented things)

Reviewer: Got it. The original sentence makes more sense

# v1.0.0 release changelogs

<!-- The following line should NOT be put in the official release changelogs -->
⚠️ Since this is a release candidate (RC), we do NOT recommend using it in a production environment. Is something not working as expected? We welcome [bug reports](https://github.com/meilisearch/meilisearch/issues/new/choose) and [feedback about new features](https://github.com/meilisearch/product/discussions).

We are proud to announce we are releasing the first major release of Meilisearch! 🎉 This new release, contrary to v0.30.0, focus on stabilization instead of bringing new features. We took the opportunity to make the final touch regarding breaking changes in the CLI and the error handler.

> To ensure the stability of Meilisearch, we will not introduce any breaking changes in future versions until the release of v2.0.0. Please refer to our [versioning policy](https://github.com/meilisearch/engine-team/blob/main/resources/versioning-policy.md) for a definition of what we consider to be a breaking change in Meilisearch.

Reviewer: ```suggestion

```
Moved this above

  * Translate multi-word synonyms into [phrases](https://docs.meilisearch.com/learn/what_is_meilisearch/features.html#phrase-search) when interpreting a search query. This change improves the relevancy of the results and stabilises the latency of search queries containing many multi-word synonyms, removing a source of denial-of-service attacks. (#3269 and meilisearch/milli#732) @loiclec
  * Improve the performance of the `proximity` ranking rule for search requests ending with short words (#3269 and meilisearch/milli#733) @loiclec
  * Limit the reindexing caused by updating settings when not needed (meilisearch/milli#706 and #3269) @GregoryConrad
* Using a single dump, you can now migrate from any old version of Meilisearch (supporting the dump feature) to the latest version. We hope to facilitate your migration to v1.0.0 to make it as smooth as possible (#3170) @dureuill
* Improve [`download-latest.sh`](https://github.com/meilisearch/meilisearch/blob/main/download-latest.sh) script: integrate apple-silicon binary (#3169) @curquiza
* Clarify error message when the DB and engine versions are incompatible (#3279) @dureuill

# Breaking changes ⚠️

> Since we are releasing v1.0.0, this is the last time we are introducing breaking changes to this release before v2.0.0.

Reviewer: ```suggestion
Since we are releasing v1.0.0, this is the last time we introduce breaking changes until v2.0.0.
```

* Rename dump command from `--dumps-dir` to `--dump-dir` (#3175) @dureuill
* Remove `--snapshot-interval-sec`. `--schedule-snapshot` now accepts an optional integer value specifying the interval in seconds (#3281) @dureuill
* The `--log-level` option and `MEILI_LOG_LEVEL` environment variable now only accept the following arguments: `ERROR`, `WARN`, `INFO`, `DEBUG`, `TRACE` and `OFF` (#3293) @loiclec
* Remove hidden but usable CLI arguments `--nb-max-chunks` and `--log-every-n` (#3305) @Kerollmops

## Revamp primary key inference

We have improved the primary key inference process when receiving documents for the first time.

If an index's primary key has not been specified, Meilisearch will attempt to infer the primary key by searching for any attributes ending with `id`, such as `puid` or` _id`.

If only one such attribute is detected, it will be designated as the index's primary key.

If multiple attributes ending with `id` are detected, Meilisearch will ask you to specify the primary key instead of choosing the first found one.

If no primary key candidates are detected, Meilisearch will throw an error as before.

Done in meilisearch/milli#752 and #3269 by @dureuill

> ## Error handler changes

Author: Also, can you confirm no async error became sync?

> ## Error handler changes

Author: I had confirmation from him directly

If only one such attribute is detected, it will be designated as the index's primary key.

If multiple attributes ending with `id` are detected, Meilisearch will ask you to specify the primary key instead of choosing the first found one.

If no primary key candidates are detected, Meilisearch will throw an error as before.

Done in meilisearch/milli#752 and #3269 by @dureuill

## Error handler changes

When Meilisearch returns an error, a `code` field is returned in the JSON object. Multiples of these unclear and vague `code`s have been changed to improve the user experience.

We consider this as a breaking change since, as a developer, it can lead to break your code if you catch one of the following `code`s in your code base:
* When calling `POST /indexes/swap-indexes`
  - `duplicate_index_found` is renamed into `invalid_swap_duplicate_index_found`
  - Sending an array of indexes not containing exactly 2 `indexUids` for a swap operation object now returns `invalid_swap_indexes` instead of a `bad_request`
* When calling `GET /indexes` and `GET /indexes/:uid`

>   - Using the `limit` parameter wrongly, Meilisearch now returns `invalid_index_limit` instead of `bad_request`

Reviewer: ```suggestion
  - Incorrect use of the `limit` parameter now returns `invalid_index_limit` instead of `bad_request`
```

>   - Using the `offset` parameter wrongly, Meilisearch now returns `invalid_index_offset` instead of `bad_request`

Reviewer: ```suggestion
  - Incorrect use of the `offset` parameter now returns `invalid_index_offset` instead of `bad_request`
```

* When calling `POST /indexes` and `PUT /indexes/:uid`
  - When `uid` is missing in the payload, Meilisearch returns `missing_index_uid` instead of `bad_request`
  - Sending a value with a different type than `string` or `null` for `primaryKey` now returns `invalid_index_primary_key` instead of `bad_request`
* When calling `GET /index/:uid/documents` and `GET /indexes/:uid/documents/:uid`

>   - Using the `fields` parameter wrongly, Meilisearch now returns `invalid_document_fields` instead of `bad_request`
>   - Using the `limit` parameter wrongly, Meilisearch now returns `invalid_document_limit` instead of `bad_request`
>   - Using the `offset` parameter wrongly, Meilisearch now returns `invalid_document_offset` instead of `bad_request`
> * When calling `POST /index/:uid/documents` and `PUT /indexes/:uid/documents`
>   - Using the `primaryKey` parameter wrongly, Meilisearch now returns ` invalid_index_primary_key` instead of `bad_request`

Reviewer: ```suggestion
  - Incorrect use of the `fields` parameter now returns `invalid_document_fields` instead of `bad_request`
  - Incorrect use of the `limit` parameter now returns `invalid_document_limit` instead of `bad_request`
  - Incorrect use of the `offset` parameter now returns `invalid_document_offset` instead of `bad_request`
* When calling `POST /index/:uid/documents` and `PUT /indexes/:uid/documents`
  - Incorrect use of the `primaryKey` parameter now returns ` invalid_index_primary_key` instead of `bad_request`
```

  - When adding a document with a wrongly formatted `_geo` field, Meilisearch now returns `invalid_document_geo_field` instead of `invalid_geo_field`[RENAME]
* When calling `GET /index/:uid/search` and `POST /indexes/:uid/search`

>   - Using the `q` parameter wrongly, Meilisearch now returns `invalid_search_q` instead of `bad_request`
>   - Using the `offset` parameter wrongly, Meilisearch now returns `invalid_search_offset` instead of `bad_request`
>   - Using the `limit` parameter wrongly, Meilisearch now returns `invalid_search_limit` instead of `bad_request`
>   - Using the `page` parameter wrongly, Meilisearch now returns `invalid_search_page` instead of `bad_request`
>   - Using the `hits_per_page` parameter wrongly, Meilisearch now returns `invalid_search_hits_per_page` instead of `bad_request`
>   - Using the `attributesToRetrieve` parameter wrongly, Meilisearch now returns `invalid_search_attributes_to_retrieve` instead of `bad_request`
>   - Using the `attributesToCrop` parameter wrongly, Meilisearch now returns `invalid_search_attributes_to_crop` instead of `bad_request`
>   - Using the `showMatchesPosition` parameter wrongly, Meilisearch now returns `invalid_search_show_matches_position` instead of `bad_request`
>   - Using the `filter` parameter wrongly, Meilisearch now returns `invalid_search_filter` instead of `bad_request`
>   - Using the `sort` parameter wrongly, Meilisearch now returns `invalid_search_sort` instead of `bad_request`
>   - Using the `facets` parameter wrongly, Meilisearch now returns `invalid_search_facets` instead of `bad_request`
>   - Using the `highlightPreTag` parameter wrongly, Meilisearch now returns `invalid_search_highlight_pre_tag` instead of `bad_request`
>   - Using the `highlightPostTag` parameter wrongly, Meilisearch now returns `invalid_search_highlight_post_tag` instead of `bad_request`
>   - Using the `matchingStrategy` parameter wrongly, Meilisearch now returns `invalid_search_matching_strategy` instead of `bad_request`

Reviewer: ```suggestion
  - Incorrect use of the `q` parameter now returns `invalid_search_q` instead of `bad_request`
  - Incorrect use of the `offset` parameter now returns `invalid_search_offset` instead of `bad_request`
  - Incorrect use of the `limit` parameter now returns `invalid_search_limit` instead of `bad_request`
  - Incorrect use of the `page` parameter now returns `invalid_search_page` instead of `bad_request`
  - Incorrect use of the `hits_per_page` parameter now returns `invalid_search_hits_per_page` instead of `bad_request`
  - Incorrect use of the `attributesToRetrieve` parameter now returns `invalid_search_attributes_to_retrieve` instead of `bad_request`
  - Incorrect use of the `attributesToCrop` parameter now returns `invalid_search_attributes_to_crop` instead of `bad_request`
  - Incorrect use of the `showMatchesPosition` parameter now returns `invalid_search_show_matches_position` instead of `bad_request`
  - Incorrect use of the `filter` parameter now returns `invalid_search_filter` instead of `bad_request`
  - Incorrect use of the `sort` parameter now returns `invalid_search_sort` instead of `bad_request`
  - Incorrect use of the `facets` parameter now returns `invalid_search_facets` instead of `bad_request`
  - Incorrect use of the `highlightPreTag` parameter now returns `invalid_search_highlight_pre_tag` instead of `bad_request`
  - Incorrect use of the `highlightPostTag` parameter now returns `invalid_search_highlight_post_tag` instead of `bad_request`
  - Incorrect use of the `matchingStrategy` parameter now returns `invalid_search_matching_strategy` instead of `bad_request`
```

Despite our focus on stabilizing Meilisearch, we also took the time to improve support for Chinese and Korean!

<!-- The following lines should ONLY be put in the official release changelogs -->
🧰 Most integrations (SDKs, clients, libraries) will be compatible with this version within four hours of the release. Sometimes this can take up to 48 hours, depending on the issues we encounter during the release.

<!-- The following lines should ONLY be put in the official release changelogs -->
Here is the exhaustive list of integrations  **not immediately compatible** with v1.0.0:
- TBD

<!-- The following line should NOT be put in the official release changelogs -->
📖 Meilisearch's official documentation will be ready for the official v1.0.0 release. Meanwhile, work-in-progress documentation is available [here]().

# New features and improvements 🔥

## Improve language support

* Add support for Korean 🇰🇷
* Improve support for Chinese:

>   * Normalize Chinese characters into Pinyin
>   * Disable the HMM (Hidden Markov Model algorithm) feature in Chinese segmentation

Reviewer 5: ```suggestion
  * Normalize Chinese characters into Pinyin
  * Disable the HMM (Hidden Markov Model algorithm) feature in Chinese segmentation
  * Unify Chinese characters variants
```

> * Improve Thai language support

Reviewer 5: ```suggestion
* Enhance Hebrew, Thai, Arabic, and Latin Languages support by improving Diacritics and nonspacing marks normalization
```

        <li>Incorrect use of the <code>offset</code> parameter now returns <code>invalid_search_offset</code> instead of <code>bad_request</code></li>
        <li>Incorrect use of the <code>limit</code> parameter now returns <code>invalid_search_limit</code> instead of <code>bad_request</code></li>
        <li>Incorrect use of the <code>page</code> parameter now returns <code>invalid_search_page</code> instead of <code>bad_request</code></li>
        <li>Incorrect use of the <code>hits_per_page</code> parameter now returns <code>invalid_search_hits_per_page</code> instead of <code>bad_request</code></li>
        <li>Incorrect use of the <code>attributesToRetrieve</code> parameter now returns <code>invalid_search_attributes_to_retrieve</code> instead of <code>bad_request</code></li>
        <li>Incorrect use of the <code>attributesToCrop</code> parameter now returns <code>invalid_search_attributes_to_crop</code> instead of <code>bad_request</code></li>
        <li>Incorrect use of the <code>showMatchesPosition</code> parameter now returns <code>invalid_search_show_matches_position</code> instead of <code>bad_request</code></li>

>         <li>Incorrect use of the <code>filter</code> parameter now returns <code>invalid_search_filter</code> instead of <code>bad_request</code></li>
>         <li>Incorrect use of the <code>sort</code> parameter now returns <code>invalid_search_sort</code> instead of <code>bad_request</code></li>

Author: ```suggestion
        <li>Incorrect use of the <code>filter</code> parameter now returns <code>invalid_search_filter</code> instead of <code>invalid_filter</code></li>
        <li>Incorrect use of the <code>sort</code> parameter now returns <code>invalid_search_sort</code> instead of <code>invalid_code</code></li>
```

    </ul>
  </details>
* When calling `GET /keys` and `GET /keys/:uid`, some error `code`s have been changed.
  <details>
    <summary>Click for details</summary>
      <ul>
        <li>Incorrect use of the <code>limit</code> parameter now returns <code>invalid_api_key_limit</code> instead of <code>bad_request</code></li>
        <li>Incorrect use of the <code>offset</code> parameter now returns <code>invalid_api_key_offset</code> instead of <code>bad_request</code></li>
      </ul>
  </details>
* When calling `POST /keys`, some error `code`s have been changed.
  <details>
    <summary>Click for details</summary>
      <ul>
        <li>If the <code>actions</code> field is missing, Meilisearch now returns <code>missing_api_key_actions</code> instead of <code>missing_parameter</code></li>
        <li>If the <code>indexes</code> field is missing, Meilisearch now returns <code>missing_api_key_indexes</code> instead of <code>missing_parameter</code></li>
        <li>If the <code>expiresAt</code> field is missing, Meilisearch now returns <code>missing_api_key_expire_at</code> instead of <code>missing_parameter</code></li>
      </ul>
  </details>

> * When calling `PUT /keys`, some error `code`s have been changed.

Author: ```suggestion
* When calling `PATCH /keys`, some error `code`s have been changed.
```

* When your machines exceeds the maximum number allowed for open files, Meilisearch returns
  - `too_many_open_files` instead of `internal` as a `code`
  - `system` instead of `internal` as a `type`
* Any Meilisearch errors with `system` type returns HTTP error code `500` (#3321) @irevoire
* Some errors are now returned synchronously
  - `invalid_index_uid`
  - `invalid_settings_ranking_rules`
  - `invalid_settings_typo_tolerance` if `oneTypo` and `twoTypos` are filled (for `minWordSizeForTypos`) but are invalid

Done by @irevoire and @loiclec in #3316, #3263, #3329, #3334 and #3339

# Fixes 🐞

* Commit SHA and date are now present in the Meilisearch logs when launching Meilisearch from a Docker image (#3212) @brunoocasali
* Fix `MDB_PAGE_FULL` internal error that would sometimes occur when inserting documents on macOS (#3210) @Kerollmops
* Re-introduce the `no_space_left_on_device` error (#3263) @irevoire
* Index creation and update dates are now kept when using a dump (#3245) @amab8901
* Fix documentation link when returning errors (#3288) @loiclec
* When using `/tasks routes`, fix the usage of `*`. More information in this issue about the bug (#3339) @loiclec

> * Fix the cancelling task behavior when sending an invalid `canceledBy` (#3375) @Kerollmops

Author: ```suggestion
* Fix the cancelling task behavior when sending an invalid `canceledBy` (#3375) @Kerollmops
* Fix dumps import for Windows, see [this issue](https://github.com/meilisearch/meilisearch/issues/2878) for more information.
```

* Remove hidden but usable CLI arguments `--nb-max-chunks` and `--log-every-n` (#3305) @Kerollmops

## Revamp primary key inference

We have improved the primary key inference process when receiving documents for the first time.

If an index's primary key has not been specified, Meilisearch will attempt to infer the primary key by searching for any attributes ending with `id`, such as `puid` or` _id`.

If only one such attribute is detected, it will be designated as the index's primary key.

If multiple attributes ending with `id` are detected, Meilisearch will ask you to specify the primary key instead of choosing the first found one.

If no primary key candidates are detected, Meilisearch will throw an error as before.

Done in meilisearch/milli#752 and #3269 by @dureuill

## Error handler changes

All detailed task responses contain `error` field. When a task fails, it is always accompanied by a JSON-formatted error response that contains an error `code` and `type`. Many of these `code`s and `type`s were unclear and have been changed to improve the user experience.

This is a breaking change if you use any of the following `code`s or `type`s in your code base:

> * When calling `POST /indexes/swap-indexes`, some error `code`s have been changed.

Reviewer: ```suggestion
* When calling `POST /indexes/swap-indexes`, the following error `code`s have been updated:
```

  <details>
    <summary>Click for details</summary>
      <ul>
        <li><code>duplicate_index_found</code> is renamed to <code>invalid_swap_duplicate_index_found</code></li>
        <li>Sending an array of indexes not containing exactly 2 `indexUids` for a swap operation object now returns <code>invalid_swap_indexes</code> instead of a <code>bad_request</code></li>
        <li>If the <code>indexes</code> field is missing, Meilisearch now returns <code>missing_swap_indexes</code> instead of <code>missing_parameter</code></li>
      </ul>
  </details>

> * When calling `GET /indexes` and `GET /indexes/:uid`, some error `code`s have been changed.

Reviewer: ```suggestion
* When calling `GET /indexes` and `GET /indexes/:uid`, the following error `code`s have been updated:
```

  <details>
    <summary>Click for details</summary>
      <ul>
        <li>Incorrect use of the <code>limit</code> parameter now returns <code>invalid_index_limit</code> instead of <code>bad_request</code></li>
        <li>Incorrect use of the <code>offset</code> parameter now returns <code>invalid_index_offset</code> instead of <code>bad_request</code></li>
    </ul>
  </details>

> * When calling `POST /indexes` and `PUT /indexes/:uid`, some error `code`s have been changed.

Reviewer: ```suggestion
* When calling `POST /indexes` and `PUT /indexes/:uid`, the following error `code`s have been updated:
```

  <details>
    <summary>Click for details</summary>
      <ul>
        <li>A missing <code>uid</code> in the payload will return <code>missing_index_uid</code> instead of <code>bad_request</code></li>
        <li>Sending a value with a other type than <code>string</code> or <code>null</code> for <code>primaryKey</code> now returns <code>invalid_index_primary_key</code> instead of <code>bad_request</code></li>
      </ul>
  </details>

> * When calling `GET /index/:uid/documents` and `GET /indexes/:uid/documents/:uid`, some error `code`s have been changed.

Reviewer: ```suggestion
* When calling `GET /index/:uid/documents` and `GET /indexes/:uid/documents/:uid`, the following error `code`s have been updated:.
```

  <details>
    <summary>Click for details</summary>
      <ul>
        <li>Incorrect use of the <code>fields</code> parameter now returns <code>invalid_document_fields</code> instead of <code>bad_request</code></li>
        <li>Incorrect use of the <code>limit</code> parameter now returns <code>invalid_document_limit</code> instead of <code>bad_request</code></li>
        <li>Incorrect use of the <code>offset</code> parameter now returns <code>invalid_document_offset</code> instead of <code>bad_request</code></li>
      </ul>
  </details>
* When calling `POST /index/:uid/documents` and `PUT /indexes/:uid/documents`
  <details>
    <summary>Click for details</summary>
      <ul>
        <li>Incorrect use of the <code>primaryKey</code> parameter now returns <code>invalid_index_primary_key</code> instead of <code>bad_request</code></li>
        <li><code>invalid_document_geo_field</code> has been renamed to <code>invalid_geo_field</code></li>
      </ul>
  </details>

> * When calling `GET /index/:uid/search` and `POST /indexes/:uid/search`, some error `code`s have been changed.

Reviewer: ```suggestion
* When calling `GET /index/:uid/search` and `POST /indexes/:uid/search`, the following error `code`s have been updated:
```

        <li>Incorrect use of the <code>showMatchesPosition</code> parameter now returns <code>invalid_search_show_matches_position</code> instead of <code>bad_request</code></li>
        <li>Incorrect use of the <code>filter</code> parameter now returns <code>invalid_search_filter</code> instead of <code>invalid_filter</code></li>
        <li>Incorrect use of the <code>sort</code> parameter now returns <code>invalid_search_sort</code> instead of <code>invalid_sort</code></li>
        <li>Incorrect use of the <code>facets</code> parameter now returns <code>invalid_search_facets</code> instead of <code>bad_request</code></li>
        <li>Incorrect use of the <code>highlightPreTag</code> parameter now returns <code>invalid_search_highlight_pre_tag</code> instead of <code>bad_request</code></li>
        <li>Incorrect use of the <code>highlightPostTag</code> parameter now returns <code>invalid_search_highlight_post_tag</code> instead of <code>bad_request</code></li>
        <li>Incorrect use of the <code>matchingStrategy</code> parameter now returns <code>invalid_search_matching_strategy</code> instead of <code>bad_request</code></li>
    </ul>
  </details>

> * When calling any `/settings` routes and their sub-routes, some error `code`s have been changed.

Reviewer: ```suggestion
* When calling any `/settings` routes and their sub-routes, the following error `code`s have been updated:
```

        <li>Incorrect use of the <code>stopWords</code> setting now returns <code>invalid_settings_stop_words</code> instead of <code>bad_request</code></li>
        <li>Incorrect use of the <code>synonyms</code> setting now returns <code>invalid_settings_synonyms</code> instead of <code>bad_request</code></li>
        <li>Incorrect use of the <code>distinctAttribute</code> setting now returns <code>invalid_settings_distinct_attribute</code> instead of <code>bad_request</code></li>
        <li>Incorrect use of the <code>typoTolerance</code> setting now returns <code>invalid_settings_typo_tolerance</code> instead of <code>bad_request</code></li>
        <li><code>invalid_typo_tolerance_min_word_size_for_typos</code> is not returned anymore and <code>invalid_settings_typo_tolerance</code> is returned instead</li>
        <li>Incorrect use of the <code>faceting</code> setting now returns <code>invalid_settings_faceting</code> instead of <code>bad_request</code></li>
        <li>Incorrect use of the <code>pagination</code> setting now returns <code>invalid_settings_pagination</code> instead of <code>bad_request</code></li>
    </ul>
  </details>

> * When calling any `/tasks` routes, some error `code`s have been changed.

Reviewer: ```suggestion
* When calling any `/tasks` routes, the following error `code`s have been updated:
```

        <li>Incorrect use of the <code>beforeEnqueuedAt</code> parameter now returns <code>invalid_task_before_enqueued_at</code> instead of <code>invalid_task_date_filter</code></li>
        <li>Incorrect use of the <code>afterEnqueuedAt</code> parameter now returns <code>invalid_task_after_enqueued_at</code> instead of <code>invalid_task_date_filter</code></li>
        <li>Incorrect use of the <code>beforeStartedAt</code> parameter now returns <code>invalid_task_before_started_at</code> instead of <code>invalid_task_date_filter</code></li>
        <li>Incorrect use of the <code>afterStartedAt</code> parameter now returns <code>invalid_task_after_started_at</code> instead of <code>invalid_task_date_filter</code></li>
        <li>Incorrect use of the <code>beforeFinishedAt</code> parameter now returns <code>invalid_task_before_finished_at</code> instead of <code>invalid_task_date_filter</code></li>
        <li>Incorrect use of the <code>afterFinishedAt</code> parameter now returns <code>invalid_task_after_finished_at</code> instead of <code>invalid_task_date_filter</code></li>
    </ul>
  </details>

> * When calling `GET /keys` and `GET /keys/:uid`, some error `code`s have been changed.

Reviewer: ```suggestion
* When calling `GET /keys` and `GET /keys/:uid`, the following error `code`s have been updated:
```

  <details>
    <summary>Click for details</summary>
      <ul>
        <li>Incorrect use of the <code>limit</code> parameter now returns <code>invalid_api_key_limit</code> instead of <code>bad_request</code></li>
        <li>Incorrect use of the <code>offset</code> parameter now returns <code>invalid_api_key_offset</code> instead of <code>bad_request</code></li>
      </ul>
  </details>

> * When calling `POST /keys`, some error `code`s have been changed.

Reviewer: ```suggestion
* When calling `POST /keys`, the following error `code`s have been updated:
```

  <details>
    <summary>Click for details</summary>
      <ul>
        <li>If the <code>actions</code> field is missing, Meilisearch now returns <code>missing_api_key_actions</code> instead of <code>missing_parameter</code></li>
        <li>If the <code>indexes</code> field is missing, Meilisearch now returns <code>missing_api_key_indexes</code> instead of <code>missing_parameter</code></li>
        <li>If the <code>expiresAt</code> field is missing, Meilisearch now returns <code>missing_api_key_expire_at</code> instead of <code>missing_parameter</code></li>
      </ul>
  </details>

> * When calling `PATCH /keys`, some error `code`s have been changed.

Reviewer: ```suggestion
* When calling `PATCH /keys`, the following error `code`s have been updated:
```

  <details>
    <summary>Click for details</summary>
      <ul>

>         <li>If trying to update the <code>uid</code> field, Meilisearch now returns <code>immutable_api_key_uid</code> instead of <code>immutable_field</code></li>
>         <li>If trying to update the <code>actions</code> field, Meilisearch now returns <code>immutable_api_key_actions</code> instead of <code>immutable_field</code></li>
>         <li>If trying to update the <code>indexes</code> field, Meilisearch now returns <code>immutable_api_key_indexes</code> instead of <code>immutable_field</code></li>
>         <li>If trying to update the <code>expiresAt</code> field, Meilisearch now returns <code>immutable_api_key_expires_at</code> instead of <code>immutable_field</code></li>
>         <li>If trying to update the <code>createdAt</code> field, Meilisearch now returns <code>immutable_api_key_created_at</code> instead of <code>immutable_field</code></li>
>         <li>If trying to update the <code>updatedAt</code> field, Meilisearch now returns <code>immutable_api_key_updated_at</code> instead of <code>immutable_field</code></li>

Reviewer: ```suggestion
        <li>Updating the <code>uid</code> field now returns <code>immutable_api_key_uid</code> instead of <code>immutable_field</code></li>
        <li>Updating the <code>actions</code> field now returns <code>immutable_api_key_actions</code> instead of <code>immutable_field</code></li>
        <li>Updating the <code>indexes</code> field now returns <code>immutable_api_key_indexes</code> instead of <code>immutable_field</code></li>
        <li>Updating the <code>expiresAt</code> field now returns <code>immutable_api_key_expires_at</code> instead of <code>immutable_field</code></li>
        <li>Updating the <code>createdAt</code> field now returns <code>immutable_api_key_created_at</code> instead of <code>immutable_field</code></li>
        <li>Updating the <code>updatedAt</code> field now returns <code>immutable_api_key_updated_at</code> instead of <code>immutable_field</code></li>
```
